R/facet-grid-.R

```diff
@@ -367,17 +407,53 @@ FacetGrid <- ggproto("FacetGrid", Facet,
       theme$panel.spacing.y %||% theme$panel.spacing)
 
     # Add axes
-    panel_table <- gtable_add_rows(panel_table, max_height(axes$x$top),     0)
-    panel_table <- gtable_add_rows(panel_table, max_height(axes$x$bottom), -1)
-    panel_table <- gtable_add_cols(panel_table, max_width(axes$y$left),     0)
-    panel_table <- gtable_add_cols(panel_table, max_width(axes$y$right),   -1)
-    panel_pos_col <- panel_cols(panel_table)
-    panel_pos_rows <- panel_rows(panel_table)
+    if (params$draw_axes$x) {
+      # Take facet_wrap approach to axis placement
+      axes$x$top <- matrix(axes$x$top[x_axis_order],
+                           nrow = nrow, ncol = ncol, byrow = TRUE)
+      panel_table <- weave_tables_row(
+        panel_table, axes$x$top, -1,
+        unit(apply(axes$x$top, 1, max_height, value_only = TRUE), "cm"),
+        name = "axis-t", z = 3
+      )
+      axes$x$bottom <- matrix(axes$x$bottom[x_axis_order],
+                              nrow = nrow, ncol = ncol, byrow = TRUE)
+      panel_table <- weave_tables_row(
+        panel_table, axes$x$bottom, 0,
+        unit(apply(axes$x$bottom, 1, max_height, value_only = TRUE), "cm"),
+        name = "axis-b", z = 3
+      )
```

Comment:
Can this be abstracted away if we are using it multiple places in the code base (both for x and y and in facet_grid and facet_wrap?

Reply:
I'd like this too, but the actual facet_wrap code has very involved code dealing with empty panels (restoring axes for non-empty panels and getting the correct axis sizes). This is right between the `matrix(axes$...)` and `weave_tables_*()` steps. We can skip all of this because facet_grid doesn't have empty panels. So while the approach is the same, it would be tricky to unify the actual code.

Reply:
could we at least abstract it away for x and y?

Reply:
I'll see what I can do

Reply:
I figured that `facet_wrap()` didn't need to measure the axis before dealing with empty panels, so axis measurement and adding it to the table is now wrapped in a `weave_axes()` helper. This helper is also used in `facet_grid()` when required.

Reply:
see... I knew you could do it